rfcs/2021-10-22-9480-processing-arrays-of-events.md

```diff
@@ -0,0 +1,262 @@
+
+The simplest way to represent an arbitrary sized array of events is the
+built-in `Vec` type. However, in the short-term, most producing
+components (sources and transforms) will only be outputting a single
+event at a time. As such, it is worth optimizing the data type for a
+single event being a common case. This is an ideal use of the `SmallVec`
+type, which can store a fixed number of elements inline, or switch to a
+standard `Vec` when that overflows.
+
+Additionally, since sources will produce arrays of exclusively logs or
+metrics, and consuming components need to detect what type of data is
+contained in the array, this type is required to provide that
+information up-front in a similar manner the base `Event` type does.
+
+```rust
+pub enum EventVec {
+    Logs(SmallVec<[LogEvent; 1]>),
+    Metrics(SmallVec<[Metric; 1]>),
+}
+```
+
+#### Generic event container trait
+
+Several of the components will need to be generic over what type of data
+they are handling, either a single `Event` or an array. This can be
+simply modelled as an iterator using existing traits. We can add
+additional methods to this trait later as needed to support needs beyond
+simple iteration (ie batching in sinks).
+
+```rust
+trait EventContainer: ByteSizeOf {
+    type EventIter;
+    type LogIter;
+    type MetricIter;
+    fn into_events(self) -> Self::EventIter;
+    fn into_logs(self) -> Self::LogIter;
+    fn into_metrics(self) -> Self::MetricIter;
+}
+
+impl IntoIterator for EventVec { … }
+impl IntoIterator for Event { … }
+```
+
+#### Enhancing the `Pipeline`
+
+The `Pipeline` structure stands as the primary unit for moving data
+between components in Vector. It receives from a source or transform,
+passes through optional inline transforms, and then sends the result
+one-by-one to a receiving transform or sink. However, before sending to
+the receiver, these events are pushed into a queue.
+
+In order to receive events asynchronously from sources, the pipeline
+implements the `Sink` trait for single events. This trait is
+parameterized over the type which the pipeline can receive. The current
+implementation allows for sending a single `Event`, and an additional
+implementation can be added for event arrays. This allows sources to be
+converted individually to send arrays while retaining compatibility with
+existing sources.
+
+```rust
+impl Sink<EventVec> for Pipeline { … }
+```
+
+#### Arrays of events in transforms
+
+There currently exist three types of transforms, expressed as traits:
+
+1. `FunctionTransform` receives a single event and outputs into a
+   mutable vector of events.
+2. `FallibleFunctionTransform` is similar, but has separate outputs for
+   successful and failed events.
+3. `TaskTransform` runs as an asynchronous task, accepting a stream of
+   events and outputting a stream of events.
+
+These will need to be rewritten to accept an event container as their
+input, and as the output for the task. While most transforms can be
```

Comment:
> I would argue that our current operation is precisely that communication cost we want to amortize. Gathering the individual events into a `Vec` on the _receiving_ side of the channel still requires that the events be sent one-by-one over the channel, which involves inter-thread communication etc. "Sending" the events at this point to the transform is simply a function call, which is hardly worth amortizing at this level.

Sounds like we agree here. The important part is changing the communication not to send events one by one over the channel, and whether or not we "amortize" the function call by pushing the loop into the transform body shouldn't matter. We're essentially optimizing out the current `ready_chunks` implementation by chunking the communication itself.

> AFAICT this type change is still required for task transforms, as the topology is doing no looping for them, so some work is still required here.

Yep, I believe this is correct.

Reply:
Ok, good. Would you like me to rework this section to skip rewriting the function transforms, or has this addressed your concerns.

Reply:
I'd prefer that we not rewrite the function-style transforms, since it's not necessary and doesn't seem to add value. We're also likely to reconsider those APIs more thoroughly in the near-ish future, so it could be duplicate work anyway.